sdks/python/apache_beam/runners/worker/operations.py

```diff
@@ -743,11 +743,36 @@ def pcollection_count_monitoring_infos(self, tag_to_pcollection_id):
     return infos
 
 
+class SdfTruncateSizedRestrictions(DoOperation):
+  def __init__(self, *args, **kwargs):
+    super(SdfTruncateSizedRestrictions, self).__init__(*args, **kwargs)
+    self.sdf_process_op = None
+
+  def current_element_progress(self):
+    # type: () -> Optional[iobase.RestrictionProgress]
+    return self.sdf_process_op.current_element_progress()
+
+  def try_split(self, fraction_of_remainder):  # type: (...) -> Optional[Any]
+    result = self.sdf_process_op.try_split(fraction_of_remainder)
+    if result is not None:
+      return result
+    return None
+
+  def add_receiver(self, operation, output_index=0):
```

Comment:
Why did we go down this path instead of replicating what DataInputOperation was doing where it accessed receivers[0] and invoked progress/split on it directly (which invokes progress/split on the ConsumerSet which only gets forwarded if it is a SingletonConsumerSet)?

This transform always has a single output so receivers[0] should always be populated.

Reply:
I did this for performance concerns. The `receivers` in python is a map of <output_index, list of receivers>. It's not efficient to get only one element from a map of <output_index, list of receivers> every time when `try_split` and `progress` are called. We can cache the receiver either when the first time `try_split` or `progress` is called, or when receivers are added.